keps/sig-network/3622-disable-lb-cluster-ip/README.md

```diff
@@ -0,0 +1,331 @@
+  - [Rollout, Upgrade and Rollback Planning](#rollout-upgrade-and-rollback-planning)
+  - [Monitoring Requirements](#monitoring-requirements)
+  - [Dependencies](#dependencies)
+  - [Scalability](#scalability)
+  - [Troubleshooting](#troubleshooting)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+  - [APIServer Flag](#apiserver-flag)
+  - [Service Type=LoadBalancerWithoutClusterIP](#service-typeloadbalancerwithoutip)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [X] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [X] (R) KEP approvers have approved the KEP status as `implementable`
+- [X] (R) Design details are appropriately documented
+- [X] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [X] (R) Graduation criteria is in place
+- [X] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Today, a Kubernetes Service Type=LoadBalancer will always allocate a clusterIP for every service. Though most implementations of Service Type=LoadBalancer do require clusterIP, there are several implementations that do not. When a large number of Type=LoadBalancer services(in one clsuter) do not need to allocate cluster IP, it will cause a waste of cluster IP resources, and the number of loadBalancer services created in this way is also limited by the number of ClusterIPs in the cluster. 
+
+This KEP proposes to add a new field to Service to opt out of clusterIP allocation for loadbalancers.
+## Motivation
+
+There are several implementations of Service Type=LoadBalancer API today that do not consume the clusterIP automatically allocated by Kubernetes. 
+
+Creating a loadbalancer service in this way brings a lot of limitations:
+* the number of load balancers is now limited to the number of available clusterIP. 
+* clusterIP are allocated for an LB even though they are not used.
+
+For clusters that have integrations for Service Type=LoadBalancer but don't require clusterIP should have the option to disable clusterIP allocation.
```

Comment:
how the other Pods in the cluster will be able to reach this Service? 
I think this will break the service discovery per example, that uses DNS for the svcname and returns the ClusterIP as A record

Reply:
1. In some business scenarios, some services do not need to be accessed by the pods in the cluster, but are accessed by requests from outside the cluster.In some cases, a large number of ClusterIPs that are not needed will be allocated.
2. In some actual production environments, the access within the cluster is also expected to be accessed through the cloud provider loadBalancer(service external ip) instead of ipvs/iptables, because the cloud  provider loadBalancer has session retention, health detection, traffic monitoring and more functions .

3. dns: If need access within the cluster, dns can configure the k8s_external plugin(dns can return the ExternalIP for service name)

Reply:
those are a lot of exceptions for justifying changing a default behavior 😅

Reply:
However, there is indeed a need for LB service not to be accessed in the cluster. In this case, it is really unnecessary to allocate clusterIP.